File: main/ctci-queue-using-two-stacks/QueueUsingTwoStacks.fs
module Queue =
    let inline make<'a> () = ([], []) : 'a list * 'a list

    let enqueue x (xs, ys) = x::xs, ys
    
    let private recharge = function
    | ([], [])  -> failwith "operation invalid on empty queue"
    | (xs, [])  -> [], List.rev xs
    | q         -> q
    
    let peek action q = // carries out the action on the front element
        let r = recharge q
        r |> snd |> List.head |> action
        r
    
    let dequeue q =
        let r = recharge q
        fst r, snd r |> List.tail

let queries =
    let interpret = function
    | [|1; x|]  -> Queue.enqueue x
    | [|2|]     -> Queue.dequeue
    | [|3|]     -> Queue.peek stdout.WriteLine
    | _         -> failwith "unrecognized query type"

    let getln = stdin.ReadLine
    let inline split (s : string) = s.Split ' '
    let nq = getln () |> int
    {1..nq} |> Seq.map (ignore >> getln >> split >> Array.map int >> interpret)

queries |> Seq.fold (|>) (Queue.make<int> ()) |> ignore

File: main/ctci-queue-using-two-stacks/queue-using-two-stacks.cpp
#include <iostream>
#include <stack>
#include <stdexcept>

namespace {
    class NoElementException : public std::runtime_error {
        using std::runtime_error::runtime_error;
    };

    template<typename T>
    class Queue {
    public:
        Queue() = default;
        Queue(const Queue<T>&) = delete;
        Queue(Queue<T>&&) = delete;
        Queue<T>& operator=(const Queue<T>&) = delete;
        Queue<T>& operator=(Queue<T>&&) = delete;

        bool empty() const;
        typename std::stack<T>::size_type size() const;

        void enqueue(const T x);
        T peek();
        T dequeue();

    private:
        std::stack<T> in {}, out {};

        void recharge();
    };

    template<typename T>
    bool Queue<T>::empty() const
    {
        return out.empty() && in.empty();
    }

    template<typename T>
    typename std::stack<T>::size_type Queue<T>::size() const
    {
        return out.size() + in.size();
    }

    template<typename T>
    void Queue<T>::enqueue(const T x)
    {
        in.push(x);
    }

    template<typename T>
    T Queue<T>::peek()
    {
        recharge();
        return out.top();
    }

    template<typename T>
    T Queue<T>::dequeue()
    {
        const auto ret = peek();
        out.pop();
        return ret;
    }

    template<typename T>
    void Queue<T>::recharge()
    {
        if (!out.empty()) return;

        if (in.empty())
            throw NoElementException{"can't access element from empty queue"};

        do {
            out.push(in.top());
            in.pop();
        }
        while (!in.empty());
    }

    int get()
    {
        auto x = 0;
        std::cin >> x;
        return x;
    }
}

int main()
{
    Queue<int> iq;

    for (auto q = get(); q > 0; --q) {
        switch (get()) {
        case 1:
            iq.enqueue(get());
            continue;

        case 2:
            iq.dequeue();
            continue;

        case 3:
            std::cout << iq.peek() << '\n';
            continue;

        default:
            throw std::invalid_argument{"unrecognized query type"};
        }
    }
}
